File: python/misc/quicksort.py
def quicksort(lst):
    if len(lst) <= 1:
        return lst
    smaller = [x for x in lst[1:] if x <= lst[0]]
    larger = [x for x in lst[1:] if x > lst[0]]
    return quicksort(smaller) + [lst[0]] + quicksort(larger)

if __name__ == '__main__':
    l = [1,5,7,3,5,9,3]
    print(l)
    print(quicksort(l))
    l = list(range(15))
    l.reverse()
    print(l)
    print(quicksort(l))
    l = [4,67,13,6,2,565,2,3,-4]
    print(l)
    print(quicksort(l))

File: cpp/algos_ds/quicksort/quicksort.cpp
#include <iostream>
#include <vector>
#include <sstream>
#include <string>
using namespace std;

vector<int> quicksort_helper(vector<int> list) {
    if(list.size() <= 1) {
        return list;
    }
    vector<int> smaller;
    vector<int> larger;
    for(int i = 1; i < list.size(); i++) {
        if(list[i] > list[0]) {
            larger.push_back(list[i]);
        }
        else {
            smaller.push_back(list[i]);
        }
    }
    vector<int> sorted;
    vector<int> sort_left = quicksort_helper(smaller);
    vector<int> sort_right = quicksort_helper(larger);
    sorted.insert(end(sorted), begin(sort_left), end(sort_left));
    sorted.push_back(list[0]);
    sorted.insert(end(sorted), begin(sort_right), end(sort_right));
    return sorted;
}

void quicksort(vector<int> &list) {
    list = quicksort_helper(list);
}

int main() {
    int num_tests;
    cin >> num_tests;
    for(int i = 1; i <= num_tests; i++) {
        string line;
        cin.ignore();
        getline(cin, line);
        istringstream istream(line);
        int n;
        vector<int> l;
        while(istream >> n) {
           l.push_back(n); 
        }
        ostringstream output;
        output << "Test " << i << ":" << ":\n";
        for(const int &v : l) {
            output << v << " "; 
        }
        output << "\n";
        quicksort(l);
        for(const int &v : l) {
            output << v << " "; 
        }
        output << "\n";
        cout << output.str();
    }
}
